### DISCUSSION TITLE
AlternateLanguageLayout.ino not working

### DISCUSSION THREAD

**Discussion by ppastor304** [NicoHood/HID#281] [Q&A]
Hi, I'm trying to build my own Spanish keyboard. First of all, I tried to run the example "AlternateLanguageLayout.ino" (last version of library 2.8.0) with the german layout and the combination KEY_7|MOD_LEFT_SHIFT doesn't work, it shows "$" instead of "/" as the comment says. I have configured the German keyboard in Windows too, but nothing changes. It doesn't work with the Spanish settings either. Has anyone experienced something like that? The full code just for an easy review:

```
#define HID_CUSTOM_LAYOUT

#define LAYOUT_GERMAN

#include "HID-Project.h"

void setup() {
  BootKeyboard.begin();
  delay(2000);
  BootKeyboard.write(KEY_7|MOD_LEFT_SHIFT); // this outputs the / symbol as per the german input method
}

void loop() {
}
```

Using Arduino Leonardo, Arduino IDE 1.8.13, HID-Project v2.8.0 library.

**Comment by NicoHood** [comment]
That is just not supported. Dont know where you've got this from.

Try `BootKeyboard.write('/');`

Or as an alternative:
```cpp
Keyboard.press(KEY_LEFT_SHIFT);
Keyboard.write(KEY_7);
Keyboard.releaseAll();
```

**Reply by ppastor304** [reply]
Last night after a lot of testing I managed to get it working using the "press" and "release" functions. I will try to implement my code with them.

My source is your file: [AlternateLanguageLayout.ino](https://github.com/NicoHood/HID/blob/master/examples/Keyboard/AlternateLanguageLayout/AlternateLanguageLayout.ino)

Thanks for the quick reply.

**Reply by NicoHood** [reply]
@SukkoPera Is this example correct? It seems it is not?

**Reply by SukkoPera** [reply]
No, as even [the docs of the original Leonardo Keyboard library](https://www.arduino.cc/en/Reference/KeyboardModifiers) say:

> For multiple key presses use Keyboard.press() 

That example should be modified so that it uses `press()`/`release()`. sorry I hadn't noticed it before.

**Reply by NicoHood** [reply]
But in this case it would make most sense to use `BootKeyboard.write('/')`, right? Cause the other solution with shift does not require any language setting at all. Or we should add both like: "That is what you used to do before, and this is how it works now"

**Comment by SukkoPera** [comment]
It depends on what we want the example to do. If we wanted it to always output a slash, regardless of the keymap in use, we would use `BootKeyboard.write('/')`, if we wanted it to always output the "other character" that is on the `7` key (i.e.: a slash in most European layouts, an Ampersand in the US, etc.), we would use `press()` twice, once for the modifier and once for the key.

On second thinking, maybe we could do both and document appropriately :).

**Reply by NicoHood** [reply]
Yeah, that is basically what I wrote before. I'd showcase both.

Without the "new" keyboard layout you'd have to manually search the keys (based on the us keyboard) and hit them, with the new layouts you can simply use the string api. We could also add [This example](https://github.com/NicoHood/HID/issues/277#issuecomment-819782324) for any character not being available on the keyboard.
